lib/dry/system/component.rb

```diff
@@ -47,49 +48,78 @@ class Component
       #   @return [Object#call] component's loader object
       attr_reader :loader
 
+      # Returns a component with a namespace and path provided from a file found within
+      # the given component dirs. If no file is found, a component is returned without
+      # these attributes.
+      #
+      # @return [Dry::System::Component]
       # @api private
-      def self.new(*args, &block)
-        cache.fetch_or_store([*args, block].hash) do
-          name, options = args
-          options = DEFAULT_OPTIONS.merge(options || EMPTY_HASH)
+      def self.locate(identifier, component_dirs, **options)
```

Comment:
Good suggestion. I've just pushed https://github.com/dry-rb/dry-system/pull/155/commits/be928c3059aa10a60b460f1a0f9158db89a9a259 – is that the kind of thing you were meaning? Are there areas where you think we could take it further?

Reply:
Also relevant is https://github.com/dry-rb/dry-system/pull/157/commits/199f37034eb00d13aad929296585464108450015 in my follow-up PR, where I continued to try and keep the options as whole hashes instead of splatted keyword args in some additional methods I added.

Reply:
yeah, like that. There's not much I can say about keywords in general, you just need to remember that `**varname` allocates a hash (inevitably). If a code path is hot you should think if this can be avoided. Also, mb ruby optimizes cases where you capture all keywords with `**options` and pass them untouched down with `**options`. But for all other cases allocation cannot be avoided. And actually it's 💯 the same for `*args`. Because of that we have [crazt code](https://github.com/dry-rb/dry-logic/blob/2e27997d0fdab90b3941e3ef71d93cf9ae62d591/lib/dry/logic/rule/interface.rb) in dry-logic.